### DISCUSSION TITLE
D3 compatibility with Phet

### DISCUSSION THREAD

**Discussion by hiiambob89** [phetsims/community#28] [Q&A] [answered]
Hey, I was trying to migrate some D3 graphs from a previous simulation into the Phet version I'm working on. I tried adding the graph to the dom using 
`this.addChild(new DOM(d3Graph))`
But this seems to throw:
`Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.`
I was wondering if any Phet simulations use D3? If so could I be linked to them so I can take a look? Otherwise, do you have any documentation for creating graphs in the Phet framework?

**Comment by jessegreenberg** [comment]
Hi @hiiambob89, can you please share a bit more code or a context where I can investigate? What is `d3Graph`? The first argument of `new DOM` should be an HTML Element. I am not aware of any PhET simulations that use D3.

We also have a library for graphs built with scenery at https://github.com/phetsims/bamboo if you are interested. you can find examples at https://github.com/phetsims/bamboo/tree/main/js/demo.

**Reply by hiiambob89** [reply]
`const refThetaGraph = document.createElement("div")
    refThetaGraph.id = "staticSim-theta";
    const d3Graph = drawTheta(model.graphData, model.graphLen, "staticSim-theta", "test");
    this.addChild(new DOM(d3Graph))`
So I create an element and pass it into my drawTheta function.
`export function drawTheta(globalData,graphLen,divID,type){
    console.log(globalThis.window.d3);
    const margin = {top: 10, right: 30, bottom: 30, left: 60},
      width = 400 - margin.left - margin.right,
      height = 400 - margin.top - margin.bottom;
  
  // append the svg object to the body of the page
  const svg = globalThis.window.d3.select(`#${divID}`)
    .append("svg")
      .attr("width", width + margin.left + margin.right)
      .attr("height", height + margin.top + margin.bottom)
    .append("g")
      .attr("transform", `translate(${margin.left},${margin.top})`);
  
  //Read the data
      const minY = globalThis.window.d3.min(globalData.data, (d) => d.theta)
      const maxY = globalThis.window.d3.max(globalData.data, (d) => d.theta)
      // Add X axis --> it is a date format
      const x = globalThis.window.d3.scaleLinear()
        .domain([0,graphLen])
        .range([ 0, width]);
      svg.append("g")
        .attr("transform", `translate(0, ${height})`)
        .call(globalThis.window.d3.axisBottom(x));
  
      // Add Y axis
      const y = globalThis.window.d3.scaleLinear()
        .domain([minY, maxY])
        .range([ height, 0 ]);
      svg.append("g")
        .call(globalThis.window.d3.axisLeft(y));
  
      
      // Add the line
      svg
        .append("path")
        .datum(globalData.data)
        .attr("fill", "none")
        .attr("stroke", "steelblue")
        .attr("stroke-width", 1.5)
        .attr("d", globalThis.window.d3.line()
          .x(function(d) { return x(d.time) })
          .y(function(d) { return y(d.theta) })
          )
      svg.append("text")
      .attr("transform", "rotate(-90)")
      .attr("x", -(height/2))
      .attr("y", -30)
      .style("text-anchor", "middle")
      .style("font-size", "20px")
      .text("θ")
  
      svg.append("text")
      .attr("transform", "translate(" + (width/2) + "," + (height + 30) + ")")
      .style("text-anchor", "middle")
      .text("Time (s)")
  
      svg.append("text")
      .attr("x", (height/2.5))
      .attr("y", 2)
      .style("text-anchor", "middle")
      .style("font-size", "15px")
      .text(`θ over time: ${type} eqn`)
    
      return svg;
  }`
Draw theta returns a d3 element. I'm assuming that the DOM constructor and the d3 element are just incompatible?
I initially tried importing d3 as mentioned in their documentation:
`import * as d3 from "https://cdn.jsdelivr.net/npm/globalThis.window.d3@7/+esm";`
but this throws a CORS error in phet. So how I did it in the example was to import it in the HTML and set it as a global. I used the minified d3 js file. For other libraries, I am using I have had success importing them from whichever source site in HTML and setting it as global as well.

**Reply by hiiambob89** [reply]
apologies for the formatting, i think I had some backtick in the code

**Reply by jessegreenberg** [reply]
Hey @hiiambob89 , just looking at some d3 examples and I noticed some functions are returning `svg.node()` like this: https://d3js.org/getting-started#d3-in-vanilla-html

I think that is the actual svg element that scenery.DOM would need. Can you please try that?

**Reply by jessegreenberg** [reply, marked as answer]
Here is a modified version of that script that seems to be working in an environment I set up. Hopefully it can be modified to work with your Display.

<details>

```js
const refThetaGraph = document.createElement( 'div' );
document.body.appendChild( refThetaGraph );
refThetaGraph.id = 'staticSim-theta';

const rootNode = new phet.scenery.Node();
const display = new phet.scenery.Display( rootNode );
document.body.appendChild( display.domElement );

// resize the display to fill the window and update it every animation frame
display.updateOnRequestAnimationFrame();
display.resizeOnWindowResize();

const d3Graph = drawTheta( {}, null, 'staticSim-theta', 'test' );
const d3GraphNode = new phet.scenery.DOM( d3Graph );
rootNode.addChild( d3GraphNode );

// Put the graph at the center bottom of the screen.
d3GraphNode.centerBottom = new phet.dot.Vector2( window.innerWidth / 2, window.innerHeight );

function drawTheta( globalData, graphLen, divID, type ) {

  // Declare the chart dimensions and margins.
  const margin = { top: 10, right: 30, bottom: 30, left: 60 };
  const width = 400 - margin.left - margin.right;
  const height = 400 - margin.top - margin.bottom;

  // Declare the x (horizontal position) scale.
  const x = d3.scaleUtc()
    .domain( [ new Date( '2023-01-01' ), new Date( '2024-01-01' ) ] )
    .range( [ margin.left, width - margin.right ] );

// Declare the y (vertical position) scale.
  const y = d3.scaleLinear()
    .domain( [ 0, 100 ] )
    .range( [ height - margin.bottom, margin.top ] );

  // append the svg object to the body of the page
  const svg = window.d3.select( `#${divID}` )
    .append( 'svg' )
    .attr( 'width', width + margin.left + margin.right )
    .attr( 'height', height + margin.top + margin.bottom );

  // Instead of these, you can position the scenery Node containing the graph.
  // .append( 'g' )
  // .attr( 'transform', `translate(${margin.left},${margin.top})` );

// Add the x-axis.
  svg.append( 'g' )
    .attr( 'transform', `translate(0,${height - margin.bottom})` )
    .call( d3.axisBottom( x ) );

// Add the y-axis.
  svg.append( 'g' )
    .attr( 'transform', `translate(${margin.left},0)` )
    .call( d3.axisLeft( y ) );

  return svg.node();
}
```
</details>

**Reply by hiiambob89** [reply]
I modified what you've posted and it works! Thanks!

**Comment by brettfiedler** [comment]
Hi @hiiambob89 ! Thanks for posting. For your notice, we're transitioning the discussions board over to https://github.com/orgs/scenerystack/discussions for a more focused space for development, distinct from our specific collection of PhET simulations. If it's no problem with you, I'm going to copy over this discussion to that forum for anyone looking in the future.

**Reply by hiiambob89** [reply]
Sounds good!
